### PR TITLE
[nix-cache] Prevent checking cache twice per package

### PR DESCRIPTION
**Pull request by mikeland73**
## Summary

Fixes bug described here https://github.com/jetify-com/devbox/pull/2054

A better (but a bit more involved) solution is to parallelize by output. Currently we parallelize by meta-output (which includes `__default_output__`). This has two downsides:

* `__default_output__` can be multiple outputs, in which case we don't parallelize them at all.
* We still do redundant work when `__default_output__` is multiple outputs. (once as `__default_output__` and once for each individual outout that forms part of defaults).

### CODE REVIEW

File: internal/devpkg/narinfo_cache.go [hunk 1]
```diff
@@ -107,13 +108,30 @@ func (p *Package) fetchNarInfoStatusOnce(output string) (bool, error) {
...
 	}
 	return f.(inCacheFunc)()
 }
 
+func (p *Package) keyForOutput(output string) string {
+	if output == useDefaultOutput {
+		sysInfo, err := p.sysInfoIfExists()
+		// let's be super safe to always avoid empty key.
+		if err == nil && sysInfo != nil && len(sysInfo.DefaultOutputs()) > 0 {
+			output = lo.Reduce(
```
Comment (gcurtis) [inline internal/devpkg/narinfo_cache.go:122]: Do the default outputs needed to be sorted to make the key stable?

**Review by gcurtis — APPROVED** [review]
I don't understand why `fetchNarInfoStatus` ended up calling twice. Did we end up getting different keys for the same installable?

File: internal/devpkg/narinfo_cache.go (hunk 1 above)
Comment (mikeland73) [reply internal/devpkg/narinfo_cache.go:122]: The implementation of `DefaultOutputs` is pretty stable, but it doesn't hurt.

File: internal/devpkg/narinfo_cache.go [hunk 2]
```diff
@@ -107,13 +108,30 @@ func (p *Package) fetchNarInfoStatusOnce(output string) (bool, error) {
...
+				sysInfo.DefaultOutputs(),
+				func(acc string, o lock.Output, _ int) string {
+					return acc + o.Name
+				},
+				"",
+			)
```
[the remark this answers is not in the document]
Comment (mikeland73) [reply internal/devpkg/narinfo_cache.go:128]: outputs are not strings, but maybe if they have a `String()` method this will work? will try

**Comment by mikeland73** [comment]
> I don't understand why fetchNarInfoStatus ended up calling twice. Did we end up getting different keys for the same installable?

yeah, we call it once per output and once for `__default_output__`. 

So for a given package with a single (default) output `out`, we call it twice, once cached under key with `out` and once with key `__default_output__`

File: internal/devpkg/narinfo_cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikeland73) [reply internal/devpkg/narinfo_cache.go:128]: it doesn't work.

**Pull request merged**